### PR TITLE
Fix: admin GUI not working with 1000s of realms

### LINKED ISSUE #31956: Admin console not usable when instance has a 1000 realms
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

When trying to access the admin UI, the endpoint to retrieve all realms is called: `/admin/realms/master/ui-ext/realms/names` which loads all realms from DB (one at a time) then checks if the user can access them (again loading from DB one at a time) which again loads the client of the realm (one at a time) to check if the user has the required role to manage/view the realm. Doing this a 1000 times take more than 60 seconds causing the TransactionReaper to intervene an close the connection.

### Version

25.0.1

### Regression

- [ ] The issue is a regression

### Expected behavior

Be able to use the Admin GUI

### Actual behavior

Unable to list Realms

### How to Reproduce?

Create a 1000 realms and open the admin console.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by yelhouti**
Search by RealmName is done before loading all realms when filtering

Closes #31956

### CODE REVIEW

**Review by edewit — DISMISSED** [review]

**Comment by yelhouti** [comment]
@edewit Thanks a lot, would I need to rebase or is it handled automatically by github or the maintainers ?

**Review by ahus1 — APPROVED** [review]

**Comment by ahus1** [comment]
@edewit - I added a default implementation and release notes. Can you please re-approve? Thanks!

**Review by sguilhen — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Review by edewit — APPROVED** [review]

File: server-spi/src/main/java/org/keycloak/models/RealmProvider.java [hunk 1]
```diff
@@ -64,6 +64,15 @@ public interface RealmProvider extends Provider {
      */
     Stream<RealmModel> getRealmsStream();
 
+    /**
+     * Returns realms as a stream filtered by search.
+     * @param search String to search for in realm names
+     * @return Stream of {@link RealmModel}. Never returns {@code null}.
+     */
+    default Stream<RealmModel> getRealmsStream(String search) {
```
Comment (indera) [inline server-spi/src/main/java/org/keycloak/models/RealmProvider.java:72]: Why did we add a less-efficient option?

Comment (indera) [reply server-spi/src/main/java/org/keycloak/models/RealmProvider.java:72]: Perhaps we should remove this or at least point to the other option 🤷

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 2]
```diff
@@ -152,6 +152,16 @@ public Stream<RealmModel> getRealmsStream() {
         return getRealms(query);
     }
 
+    @Override
+    public Stream<RealmModel> getRealmsStream(String search) {
```
Comment (indera) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:156]: We missed the opportunity to add an optional `limit` parameter, to fetch only N records

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 3]
```diff
@@ -152,6 +152,16 @@ public Stream<RealmModel> getRealmsStream() {
...
+        if (search.trim().isEmpty()) {
+            return getRealmsStream();
+        }
+        TypedQuery<String> query = em.createNamedQuery("getRealmIdsWithNameContaining", String.class);
+        query.setParameter("search", search);
```
Comment (indera) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:161]: Is this worth it?

```
query.setParameter("limit", limit);
query.setParameter("offset", offset);

```

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 2 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:156]: The result is a stream, so limiting the result is an implicit functionality of streams.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 3 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:161]: This PR is closed. If you think there is the need to optimize it, feel free to open an issue and a PR.

File: server-spi/src/main/java/org/keycloak/models/RealmProvider.java (hunk 1 above)
Comment (indera) [reply server-spi/src/main/java/org/keycloak/models/RealmProvider.java:72]: I just have to say it, this can make keycloak much more desirable to use!

**Pull request merged**